Support for :perl, :perlfile, :perldo and perleval()

jacquesg opened the pull request:
Initial support for ```:perl```, ```:perlfile```, ```:perldo``` and ```perleval()```. These have been implemented via the perl remote plugin infrastructure and requires [```Neovim::Ext```](https://metacpan.org/release/JACQUESG/Neovim-Ext-0.03) version 0.03 or greater.

jacquesg committed:
support for :perl, :perlfile, :perldo and perleval()

janlazo commented on test/functional/provider/perl_spec.lua:
```diff
@@ -19,7 +23,51 @@ before_each(function()
 
-describe('perl host', function()
+describe('legacy perl provider', function()
+  if helpers.pending_win32(pending) then return end
```
Which tests fail on Windows?

jacquesg replied:
The remote plugin provider doesn't work on Windows currently as its not possible to poll STDIN yet.

jacquesg replied:
If or when ```Neovim::Ext``` can use a libuv based backend maybe this would start working. This isn't a new limitation.

justinmk commented on src/nvim/eval/funcs.c:
```diff
@@ -6374,6 +6374,14 @@ static void f_pyxeval(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
+/*
+ * "perleval()" function
+ */
```
lint is failing, use `///` instead

jacquesg replied:
Fixed.

justinmk commented:
lgtm but travis is failing. 

thanks for doing this!

jacquesg commented:
> lgtm but travis is failing.
> 
> thanks for doing this!

They're passing now, when I submitted the PR the new version of ```Neovim::Ext``` wasn't index yet, so it still downloaded the older version.

There seems to be 2 memory leaks, which I cannot explain, given that I based this 100% off the existing python code.

janlazo commented:
Can you update the documentation (maybe `:h provider-perl`) that this does not work on Windows https://github.com/neovim/neovim/pull/12809#discussion_r479746116?

janlazo commented on runtime/doc/index.txt:
```diff
@@ -1441,6 +1441,9 @@ tag		command		action ~
 |:packloadall|	:packl[oadall]	load all packages under 'packpath'
 |:pclose|	:pc[lose]	close preview window
 |:pedit|	:ped[it]	edit file in the preview window
+|:perl|		:perl		execute perl command
```
Extra tab?

jacquesg replied:
Thought so as well, but it's just the way I think github renders it.

jacquesg committed:
document that the perl provider doesn't currently work on Windows

jacquesg committed:
document that the perl legacy interface is now also supported

jacquesg commented:
> #12809 (comment)

Done. I've just realised that the original compiled in version in vim had a ```VIM``` package which was compiled into vim. These are not currently available, but I think most if not all of them can be provided by adding a compatibility layer. I cannot simply just add a ```VIM``` package and publish it on CPAN, I guess I'd need to namespace it properly and do a runtime fixup.

justinmk commented:
> There seems to be 2 memory leaks, which I cannot explain, given that I based this 100% off the existing python code.

yeah that is strange. the rettv returned by `f_rpcrequest` should be handled like any other Vimscript function. And why is it only 1 byte leaked?

I wonder what the contents of `result` are here: https://github.com/neovim/neovim/blob/c6648fe8820003a92e034e7717a86c1e0616ffb0/src/nvim/eval/funcs.c#L7339

Can you try printing that `result` ? something like (untested):

      ELOG("xxxxxxx %s", result.string.data);


```
[32m[2m[ RUN      ][0m[0m legacy perl provider :perlfile command: [2m146.40 ms[0m [1m[32mOK[0m[0m
= ==23145==ERROR: LeakSanitizer: detected memory leaks
= Direct leak of 1 byte(s) in 1 object(s) allocated from:
=     #0 0x502cc3 in malloc /tmp/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
...
=     #4 0x118ea3e in xmemdupz /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:200:17
=     #5 0x6496d7 in object_to_vim /home/travis/build/neovim/neovim/build/../src/nvim/api/private/helpers.c:1102:29
=     #6 0xa873c4 in f_rpcrequest /home/travis/build/neovim/neovim/build/../src/nvim/eval/funcs.c:7377:8
...
=     #19 0x8f2e62 in eval0 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3121:9
...
=     #25 0x971d3c in eval_call_provider /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:10428:9
=     #26 0xc16ea8 in script_host_execute /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:4175:11
=     #27 0xc17b5e in ex_perl /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:940:3
=     #28 0xc70c35 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2251:5
=     #29 0xc59667 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:615:20
= 
= SUMMARY: AddressSanitizer: 1 byte(s) leaked in 1 allocation(s).
```

jacquesg commented:
> > There seems to be 2 memory leaks, which I cannot explain, given that I based this 100% off the existing python code.
> 
> yeah that is strange. the rettv returned by `f_rpcrequest` should be handled like any other Vimscript function. And why is it only 1 byte leaked?
> 
> I wonder what the contents of `result` are here:
> 
> https://github.com/neovim/neovim/blob/c6648fe8820003a92e034e7717a86c1e0616ffb0/src/nvim/eval/funcs.c#L7339
> 
> * Can you try printing that `result` ? something like (untested):
>   ```
>       ELOG("xxxxxxx %s", result.string.data);
>   ```
> * or may need to inspect it in gdb, to see what's really going on
> 
> ```
> �[32m�[2m[ RUN      ]�[0m�[0m legacy perl provider :perlfile command: �[2m146.40 ms�[0m �[1m�[32mOK�[0m�[0m
> = ==23145==ERROR: LeakSanitizer: detected memory leaks
> = Direct leak of 1 byte(s) in 1 object(s) allocated from:
> =     #0 0x502cc3 in malloc /tmp/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
> ...
> =     #4 0x118ea3e in xmemdupz /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:200:17
> =     #5 0x6496d7 in object_to_vim /home/travis/build/neovim/neovim/build/../src/nvim/api/private/helpers.c:1102:29
> =     #6 0xa873c4 in f_rpcrequest /home/travis/build/neovim/neovim/build/../src/nvim/eval/funcs.c:7377:8
> ...
> =     #19 0x8f2e62 in eval0 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3121:9
> ...
> =     #25 0x971d3c in eval_call_provider /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:10428:9
> =     #26 0xc16ea8 in script_host_execute /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:4175:11
> =     #27 0xc17b5e in ex_perl /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:940:3
> =     #28 0xc70c35 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2251:5
> =     #29 0xc59667 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:615:20
> = 
> = SUMMARY: AddressSanitizer: 1 byte(s) leaked in 1 allocation(s).
> ```

This is what is printed out:

less /Users/jacquesg/github/neovim/build/.nvimlog | grep xxxx
ERROR 2020-08-30T18:56:35.875 77763 f_rpcrequest:7348: xxxxxxx ok
ERROR 2020-08-30T18:56:35.876 77763 f_rpcrequest:7348: xxxxxxx
ERROR 2020-08-30T18:56:36.007 77766 f_rpcrequest:7348: xxxxxxx ok
ERROR 2020-08-30T18:56:36.009 77766 f_rpcrequest:7348: xxxxxxx
ERROR 2020-08-30T18:56:36.144 77769 f_rpcrequest:7348: xxxxxxx ok
ERROR 2020-08-30T18:56:36.147 77769 f_rpcrequest:7348: xxxxxxx (null)
ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx ok
ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx

justinmk commented:
> ERROR 2020-08-30T18:56:36.144 77769 f_rpcrequest:7348: xxxxxxx ok
> ERROR 2020-08-30T18:56:36.147 77769 f_rpcrequest:7348: xxxxxxx (null)
> ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx ok
> ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx

the blank cases might be worth inspecting in gdb. Could you try always emitting something like "ok" from the host, to ensure there are never blank cases, and push it here, then see if it leaks (assuming you don't want to run the leak santizier locally, though instructions are here: https://github.com/neovim/neovim/blob/master/src/nvim/README.md#build-with-asan )

jacquesg commented:
> > ERROR 2020-08-30T18:56:36.144 77769 f_rpcrequest:7348: xxxxxxx ok
> > ERROR 2020-08-30T18:56:36.147 77769 f_rpcrequest:7348: xxxxxxx (null)
> > ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx ok
> > ERROR 2020-08-30T18:56:36.279 77772 f_rpcrequest:7348: xxxxxxx
> 
> the blank cases might be worth inspecting in gdb. Could you try always emitting something like "ok" from the host, to ensure there are never blank cases, and push it here, then see if it leaks (assuming you don't want to run the leak santizier locally, though instructions are here: https://github.com/neovim/neovim/blob/master/src/nvim/README.md#build-with-asan )

I've been trying to get ASAN to work, but looks like a dead-end on macOS at this stage. I think I understand what is happening here, my suspicion is that because perl returns the last value of the last statement, this is getting sent back to Neovim, but we're not interested in the value, and we actually don't clean it up.

![image](https://user-images.githubusercontent.com/3534453/91667173-5bbc1d80-eafa-11ea-8c1d-336eb43e6016.png)

I will confirm though...

jacquesg commented:
Confirmed. I've updated Neovim-Ext to not (accidentally) return anything over RPC. I believe ```eval_call_provider```'s result shouldn't just be blindly casted to ```void``` as we do here, Neovim shouldn't leak due to something that happens outside the control of it.

I could look into this as a separate PR?

This PR I believe is now complete.

jacquesg committed:
added perleval() documentation

jacquesg committed:
import legacy perl interface tests

jacquesg committed:
only perl 5.22+ is supported

jacquesg commented:
I've found some "old" tests for vim's perl interface which I've imported and adjusted appropriately.

janlazo commented:
Blob is not supported yet.

jacquesg commented:
> Blob is not supported yet.

Ah, I think I'll take that commit out then.

janlazo commented on runtime/doc/provider.txt:
```diff
@@ -135,6 +135,7 @@ currently. The Vim legacy |perl-vim| interface is also supported (which is
 itself implemented as a Nvim remote-plugin).
 https://github.com/jacquesg/p5-Neovim-Ext
 
+Note: Only perl versions from 5.22 onward are supported.
```
Make this is an explicit requirement and update the health check.

jacquesg replied:
Done.

janlazo commented on runtime/autoload/health/provider.vim:
```diff
@@ -703,6 +703,13 @@ function! s:check_perl() abort
   endif
 
+  call s:system(['perl', '-e', 'use v5.20'])
+  if v:shell_error
```
I meant that you should update the warning above instead of adding another one.

jacquesg replied:
Ok, I've reworked this a bit more. I've made an additional change here to improve the health check. We were not honouring the ```g:perl_host_prog``` variable when we were doing the health checks, instead we were just running ```perl```.

janlazo commented on runtime/autoload/provider/perl.vim:
```diff
@@ -11,6 +11,12 @@ function! provider#perl#Detect() abort
   endif
 
+  " if perl is available, make sure we have 5.20+
+  call system([prog, '-e', 'use v5.20'])
```
Why is it 5.20+ here but the documentation requires 5.22+?

jacquesg replied:
Yep, thats not right. I've fixed it.

jacquesg committed:
healtcheck: use g:perl_host_prog if its set instead

using just 'perl' isn't correct as it may not be the version requested.
ditto for 'cpanm', rather go through 'App::cpanminus' to find the latest
perl version

justinmk commented on test/functional/provider/perl_spec.lua:
```diff
@@ -19,7 +23,51 @@ before_each(function()
   clear()
```
seems like the `missing_provider()` check (line 15) is not working, or the `provider#perl#Detect()` function is not working, since the tests are failing on the BSD CIs : 

```
[ RUN      ] legacy perl provider :perl command: 32.24 ms ERR
test/functional/helpers.lua:106: Vim(echoerr):"Neovim::Ext" cpan module is not installed

stack traceback:
	test/functional/helpers.lua:106: in function 'request'
	test/functional/helpers.lua:270: in function 'command'
	test/functional/provider/perl_spec.lua:34: in function <test/functional/provider/perl_spec.lua:33>

[ RUN      ] legacy perl provider :perlfile command: 34.05 ms ERR
test/functional/helpers.lua:106: Vim(echoerr):"Neovim::Ext" cpan module is not installed

stack traceback:
	test/functional/helpers.lua:106: in function 'request'
	test/functional/helpers.lua:270: in function 'command'
	test/functional/provider/perl_spec.lua:41: in function <test/functional/provider/perl_spec.lua:38>

justinmk replied:
i think this patch is needed : 

```diff
diff --git a/test/functional/helpers.lua b/test/functional/helpers.lua
index e8435cd3b796..5864cecb1377 100644
--- a/test/functional/helpers.lua
+++ b/test/functional/helpers.lua
@@ -768,10 +768,10 @@ function module.new_pipename()
 end
 
 function module.missing_provider(provider)
-  if provider == 'ruby' or provider == 'node' or provider == 'perl' then
+  if provider == 'ruby' or provider == 'node' then
     local prog = module.funcs['provider#' .. provider .. '#Detect']()
     return prog == '' and (provider .. ' not detected') or false
-  elseif provider == 'python' or provider == 'python3' then
+  elseif provider == 'python' or provider == 'python3' or provider == 'perl' then
     local py_major_version = (provider == 'python3' and 3 or 2)
     local errors = module.funcs['provider#pythonx#Detect'](py_major_version)[2]
     return errors ~= '' and errors or false

justinmk commented on runtime/doc/provider.txt:
```diff
@@ -129,9 +129,13 @@ To use the RVM "system" Ruby installation: >
 
-Nvim supports Perl |remote-plugin|s.
+Nvim supports Perl |remote-plugin|s on Unix platforms. Support for polling
+STDIN on MS-Windows in lacking from all known event loop implementations
```
```suggestion
STDIN on MS-Windows is lacking from all known event loop implementations
```

jacquesg replied:
Fixed.

justinmk commented:
LGTM, just need to fix `missing_provider()` : 

```diff
diff --git a/test/functional/helpers.lua b/test/functional/helpers.lua
index e8435cd3b796..53b33de59f7c 100644
--- a/test/functional/helpers.lua
+++ b/test/functional/helpers.lua
@@ -768,10 +768,13 @@ function module.new_pipename()
 end
 
 function module.missing_provider(provider)
-  if provider == 'ruby' or provider == 'node' or provider == 'perl' then
+  if provider == 'ruby' or provider == 'node' then
     local prog = module.funcs['provider#' .. provider .. '#Detect']()
     return prog == '' and (provider .. ' not detected') or false
-  elseif provider == 'python' or provider == 'python3' then
+  elseif provider == 'perl' then
+    local errors = module.funcs['provider#'..provider..'#Detect']()[2]
+    return errors ~= '' and errors or false
+  elseif provider == 'python' or provider == 'python3' or provider == 'perl' then
     local py_major_version = (provider == 'python3' and 3 or 2)
     local errors = module.funcs['provider#pythonx#Detect'](py_major_version)[2]
     return errors ~= '' and errors or false
```

The node/ruby cases really should be cleaned up, but meanwhile we need another special case :(   The perl provider followed the pythonx pattern, which is correct.

jacquesg commented:
> LGTM, just need to fix `missing_provider()` :
> 
> ```diff
> diff --git a/test/functional/helpers.lua b/test/functional/helpers.lua
> index e8435cd3b796..82b6fb42e848 100644
> --- a/test/functional/helpers.lua
> +++ b/test/functional/helpers.lua
> @@ -768,9 +768,12 @@ function module.new_pipename()
>  end
>  
>  function module.missing_provider(provider)
> -  if provider == 'ruby' or provider == 'node' or provider == 'perl' then
> +  if provider == 'ruby' or provider == 'node' then
>      local prog = module.funcs['provider#' .. provider .. '#Detect']()
>      return prog == '' and (provider .. ' not detected') or false
> +  elseif provider == 'perl' then
> +    local errors = module.funcs['provider#'..provider..'#Detect']()[2]
> +    return errors ~= '' and errors or false
>    elseif provider == 'python' or provider == 'python3' then
>      local py_major_version = (provider == 'python3' and 3 or 2)
>      local errors = module.funcs['provider#pythonx#Detect'](py_major_version)[2]
> ```
> 
> The node/ruby cases really should be cleaned up, but meanwhile we need another special case :( The perl provider followed the pythonx pattern, which is correct.
> 
> * edit: #12839 will help

Fixed as well.
